atom/browser/net/atom_url_loader_factory.cc

```diff
@@ -341,21 +363,32 @@ void AtomURLLoaderFactory::StartLoadingStream(
 void AtomURLLoaderFactory::SendContents(
     network::mojom::URLLoaderClientPtr client,
     network::ResourceResponseHead head,
-    const char* data,
-    size_t ssize) {
-  uint32_t size = base::saturated_cast<uint32_t>(ssize);
-  mojo::DataPipe pipe(size);
-  MojoResult result =
-      pipe.producer_handle->WriteData(data, &size, MOJO_WRITE_DATA_FLAG_NONE);
-  if (result != MOJO_RESULT_OK || size < ssize) {
-    client->OnComplete(network::URLLoaderCompletionStatus(net::ERR_FAILED));
+    std::string data) {
+  head.headers->AddHeader(kCORSHeader);
+  client->OnReceiveResponse(head);
+
+  // Code bellow follows the pattern of data_url_loader_factory.cc.
+  mojo::ScopedDataPipeProducerHandle producer;
+  mojo::ScopedDataPipeConsumerHandle consumer;
+  if (mojo::CreateDataPipe(nullptr, &producer, &consumer) != MOJO_RESULT_OK) {
+    client->OnComplete(
+        network::URLLoaderCompletionStatus(net::ERR_INSUFFICIENT_RESOURCES));
     return;
   }
 
-  head.headers->AddHeader(kCORSHeader);
-  client->OnReceiveResponse(head);
-  client->OnStartLoadingResponseBody(std::move(pipe.consumer_handle));
-  client->OnComplete(network::URLLoaderCompletionStatus(net::OK));
+  client->OnStartLoadingResponseBody(std::move(consumer));
+
+  auto write_data = std::make_unique<WriteData>();
```

Comment:
Can we simplify the self ownership of this struct `WriteData` ? I am thinking maintaining a map of `client` and `WriteData` in `AtomURLLoaderFactory` to manage the lifetime. Thoughts ?

Reply:
This pattern is giving the ownership of `WriteData` to the callback of `producer->Write`, and it destructs automatically when the write is done, so we don't need to do manual memory management.

The code here is copied from `DataURLLoaderFactory`.